### ISSUE TITLE
Jumbo frames support on ls1024a

### ISSUE THREAD

**Issue by niggowai** [scpcom/linux#6]
Hi,
it would be nice to have support for Jumbo Frames on the **ls1024a**. The stock kernel allows setting a mtu to 9000 (maybe even higher). However, the kernel here fails to set a mtu>1500. Using a higher mtu may reduce the load on the nas, since a nfs read is usually around 8k, what would fit in a 9k jumbo frame. With a 1500 mtu the nas would use 6 frames, so it also has to calculate 6 crc's instead of 1.
```shell
nas542:~# ip link set dev eth1 mtu 9000
[ 1268.740000] eth1: mtu greater than device maximum
ip: ioctl 0x8922 failed: Invalid argument
```

**Comment by scpcom** [comment]
Thank you for reporting. I fixed this now:
https://github.com/scpcom/linux/commit/04ebf3001e65271f9c6fce8c5e64335adee56149

Now it works with:
```
ip link set dev eth1 down
ip link set dev eth1 mtu 9000
ip link set dev eth1 up
```

**Comment by niggowai** [comment]
Nice, will try out this evening,
huge thanks already

**Comment by a-st** [comment]
Just tried it on my NAS542 but sadly without success
```
root@nas542:~# uname -a
Linux nas542 6.6.62+nas5xx #nas5xx SMP Fri Jan 17 16:37:36 UTC 2025 armv7l GNU/Linux

root@nas542:~# ip link set dev eth1 down

root@nas542:~# ip link set dev eth1 mtu 9000
Error: mtu greater than device maximum.

root@nas542:~# ip link set dev eth1 up

root@nas542:~# ip link show eth1
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN mode DEFAULT group default qlen 1000
    link/ether bc:99:11:5b:80:f6 brd ff:ff:ff:ff:ff:ff permaddr 00:2a:2b:2c:2d:2e
```

**Comment by scpcom** [comment]
Sorry, maybe you do have the old version of the zip file, I did overwrite it on 2025-01-20 21:11
https://seafile.servator.de/nas/zyxel/kernel/linux-image-6.6.62-20250116-nas5xx-armhf.zip
Or just wait for the next releases this week (6.1.126 and 6.6.72) than you can be sure it is included :-)

**Comment by niggowai** [comment]
Well done, i got it working by compiling from source.
I will report back if the jumbo frames give a performance uplift for nfs (need just some time to setup Jumbo on my server)

**Comment by niggowai** [comment]
Ok I finally found time to do a bit of benchmarking.

TLDR, sadly the jumbo frames do not provide a better performance than the default 1500 mtu...
however after tweaking some sysctls, i got a decent 55MB read, 35MB write with cp.
The main problem here is the cpu, a write or read consumes nearly 100%.

The tests i made wer done with rsync and cp, with the following versions:

- rsync  version 3.3.0  protocol version 31
- cp (GNU coreutils) 9.3

The NAS runs alpine linux with
```
Linux nas542 6.6.62-g04ebf3001e65 #14 SMP Fri Feb 14 17:23:25 UTC 2025 armv7l Linux
```

My Server uses fedora linux 39
```
Linux asahi 6.11.8-400.asahi.fc39.aarch64+16k #1 SMP PREEMPT_DYNAMIC Sat Nov 16 22:00:34 UTC 2024 aarch64 GNU/Linux
```

The filesystem is mounted via NFSv4.2 over a dedicated network.
```shell
nas542:/mnt/RAID /mnt/RAID nfs4 rw,relatime,vers=4.2,rsize=131072,wsize=131072,namlen=255,hard,proto=tcp,timeo=600,retrans=2,sec=sys,clientaddr=192.168.188.1,local_lock=none,addr=192.168.188.2 0 0
```

my tweaked `sysctl.conf` looks like

```
net.core.wmem_max=12582912
net.core.rmem_max=12582912
net.ipv4.tcp_rmem= 10240 87380 12582912
net.ipv4.tcp_wmem= 10240 87380 12582912
net.ipv4.tcp_no_metrics_save = 1
net.core.netdev_max_backlog = 5000
```

## Benchmarks with Jumbo frames disabled

### MTU 1500 server -> nas

```shell
┌[root@asahi] [/dev/pts/0] [1]
└[~]> rsync -ah --progress file.bin /mnt/RAID/ 
sending incremental file list
file.bin
         10,74G 100%   35,97MB/s    0:04:44 (xfr#1, to-chk=0/1)

┌[root@asahi] [/dev/pts/0] 
└[~]> time cp file.bin /mnt/RAID/ 
cp file.bin /mnt/RAID/  0,01s user 8,97s system 3% cpu 4:57,85 total
```

> Around 33MB/s for the cp command

### MTU 1500 nas -> server

```shell
┌[root@asahi] [/dev/pts/0] 
└[~]> rsync -ah --progress /mnt/RAID/file.bin .
sending incremental file list
file.bin
         10,74G 100%   50,89MB/s    0:03:21 (xfr#1, to-chk=0/1)

┌[root@asahi] [/dev/pts/0] 
└[~]> time cp /mnt/RAID/file.bin .
cp /mnt/RAID/file.bin .  0,06s user 11,27s system 5% cpu 3:18,11 total
```

> Around 50MB/s for the cp command

## Benchmarks with Jumbo frames enabled

### MTU 9000 server -> nas

```shell
┌[root@asahi] [/dev/pts/0] 
└[~]> 
sending incremental file list
file.bin
         10,74G 100%   43,04MB/s    0:03:57 (xfr#1, to-chk=0/1)

┌[root@asahi] [/dev/pts/0] 
└[~]> time cp file.bin /mnt/RAID/      
cp file.bin /mnt/RAID/  0,03s user 9,29s system 3% cpu 4:12,49 total
```

> Around 39,68MB/s for the cp command


### MTU 9000 nas -> server

```shell
┌[root@asahi] [/dev/pts/0] 
└[~]> rsync -ah --progress /mnt/RAID/file.bin .
sending incremental file list
file.bin
         10,74G 100%   34,07MB/s    0:05:00 (xfr#1, to-chk=0/1)

┌[root@asahi] [/dev/pts/0] 
└[~]> time cp /mnt/RAID/file.bin .
cp /mnt/RAID/file.bin .  0,07s user 11,11s system 4% cpu 4:10,59 total
```

> Around 40MB/s for the cp command

**Comment by scpcom** [comment]
Thank you for sharing the results. I will try the sysctls on my setup.
